Foundation/NSLock.swift

```diff
@@ -258,3 +286,80 @@ open class NSCondition: NSObject, NSLocking {
     
     open var name: String?
 }
+
+private func timeSpecFrom(date: Date) -> timespec? {
+    let interval = date.timeIntervalSince1970
+    guard interval >= 0 else {
+        return nil
+    }
+
+    var ts = timespec()
+    ts.tv_sec = Int(interval.rounded(.down))
+    ts.tv_nsec = Int((interval - Double(ts.tv_sec)) * 1_000_000_000.0)
+    return ts
+}
+
+#if os(OSX) || os(iOS)
+
+private func allocateTimedLockData() -> UnsafeMutablePointer<_timed_lock_data>? {
+    let tld = UnsafeMutablePointer<_timed_lock_data>.allocate(capacity: 1)
+    if pthread_mutex_init(&tld.pointee.mutex, nil) == 0 {
+        if pthread_cond_init(&tld.pointee.cond, nil) == 0 {
+            return tld
+        } else {
+            pthread_mutex_destroy(&tld.pointee.mutex)
+        }
+    }
+    tld.deallocate(capacity: 1)
+    return nil
+}
+
+private func deallocateTimedLockData(_ tld: UnsafeMutablePointer<_timed_lock_data>?) {
+    if let tld = tld {
+        pthread_mutex_destroy(&tld.pointee.mutex)
+        pthread_cond_destroy(&tld.pointee.cond)
+        tld.deallocate(capacity: 1)
+    }
+}
+
+private func getOrAllocateTimedLockData(timedLockData: inout UnsafeMutablePointer<_timed_lock_data>?) -> UnsafeMutablePointer<_timed_lock_data> {
+    if timedLockData == nil, let tld = allocateTimedLockData() {
+        var expected: UnsafeMutablePointer<_timed_lock_data>? = nil
+
+        // Use an atomic compare and exchange to set the pointer to the just allocated
+        // data. On X86 this is a LOCK CMPXCHG.
+        // `object` is the target to update, timedLockData, shared by all threads.
+        // `expected` is the value the target should still be (nil).
+        // `desired` is the new value to set it to.
+        if !_stdlib_atomicCompareExchangeStrongPtr(object: &timedLockData,
+                                                  expected: &expected,
+                                                  desired: tld) {
+            // The update failed. Another thread set the `object` to its allocated data
+            // before this thread so deallocate the timedLockData that isnt needed anymore.
+            // timedLockData now points to the data allocated by the other thread.
+            deallocateTimedLockData(tld)
+        }
+    }
+    return timedLockData!
+}
+
+// Emulate pthread_mutex_timedlock using pthread_cond_timedwait.
+// lock(before:) has obtained a shared pointer to a _time_lock_data structure
+// and the condition variable to be waited on.
+// unlock() will check for the existance of this structure and if its pointer is
+// non-nil, it will use pthread_cond_broadcast() to stop any waits in progress.
+private func timedLock(mutex: _PthreadMutexPointer,
+                       endTime: inout timespec,
+                       using tld: UnsafeMutablePointer<_timed_lock_data>) -> Bool {
+    pthread_mutex_lock(&tld.pointee.mutex)
```

Comment:
again I am not certain this is safe

Reply:
@phausler Who would be able to give a definitive answer as to whether it is safe or not?

Reply:
Formally it isn't guaranteed to be safe, but I can't think of anything better today.

Reply:
I think we'll need to implement it in C and call out from Swift.

Reply:
The performance may not be great, but that is a secondary concern to correctness.

Reply:
At least doing it in C saves adding a 2nd ptr to every `Lock` allocation which will mostly be wasted (except for the locks using the `lock(before:)` method. I will look into it.

Reply:
A `static inline` C function to project the pointer to the mutex field from a pointer to the struct could also still be inlined away.